### ISSUE TITLE
Feedback to plugin support documentation

### ISSUE THREAD

**Issue by zaantar** [versionpress/versionpress#1155]
Few ideas about [Plugin-Support.md](https://github.com/versionpress/versionpress/blob/8f585b1635306fc95395f00e5b451c303ffe509a/docs/Plugin-Support.md)

- In general, examples should be illustrated in yml file excerpts, not only discussed (for example `theme/switch` versus `option/edit` actions)
- Action detection: It's not entirely clear to me who (and when) should report them to VersionPress
- Files to commit with an action: In the provided examples, it's not always entirely clear what is a string literal and what is an example value (I'd prefer `$example_value` over `'example_value'`)
- VPID: Why is it needed? The reader (not too familiar with VersionPress) might get confused here, the difference between an ID and VPID is not really clear; a very brief explanation or a link to another document would help a lot here.
- References: There's a typo - "needs to understands".
- M:N references: Might be better to first show an example and _then_ describe how it works and why is it needed
- Shortcodes: It's not clear if callbacks are supported here as well.

More or less related to plugin support, three feature suggestions:

1. Add a filter to determine the plugin descriptors directory. As a plugin developer I would prefer to bury them deeper within the plugin's directory structure and avoid having a hidden directory there.
2. Ignore git repositories when VersionPress is activated (and later on). Sure, it is possible to achieve this manually but I have to admint it's quite annoying.
3. The possibility to favourite certain commits and then display only those favourites would be great. Use cases:
    - prototyping and frequently coming back to one "starting point"
    - repeatedly debugging some complex operation that involves database updates

**Comment by borekb** [comment]
@zaantar thanks for your feedback! I've made some changes that you can see in PR https://github.com/versionpress/versionpress/pull/1157. Replies are below:

>  In general, examples should be illustrated in yml file excerpts, not only discussed (for example `theme/switch` versus `option/edit` actions)

Agreed. You mention action priorities which is a relatively deep concept and certainly could use an example. At the same time, we try to keep the document relatively short so I guess we could have a footnote section or something for things like this.

- [ ] Add a footnote section with extended examples.

> Action detection: It's not entirely clear to me who (and when) should report them to VersionPress

Seems pretty clear to me from the [Actions detection](https://github.com/versionpress/versionpress/blob/49fcbc6bfa7ed8a4972974a32dec22e5ce38ab4d/docs/Plugin-Support.md#action-detection) section. What do you find unclear? Any suggestions of how to improve it?

> Files to commit with an action: In the provided examples, it's not always entirely clear what is a string literal and what is an example value (I'd prefer `$example_value` over `'example_value'`)

Agreed, I've updated the section.

> VPID: Why is it needed? The reader (not too familiar with VersionPress) might get confused here, the difference between an ID and VPID is not really clear; a very brief explanation or a link to another document would help a lot here.

I can see how the distinction between an ID and a VPID might be slightly confusing. An attempt to explain it is in the [Identifying entities](https://github.com/versionpress/versionpress/blob/49fcbc6bfa7ed8a4972974a32dec22e5ce38ab4d/docs/Plugin-Support.md#identifying-entities) section, do you have any suggestions on how to make it better?

> References: There's a typo - "needs to understands".

:+1:, fixed.

> M:N references: Might be better to first show an example and _then_ describe how it works and why is it needed

M:N references start with an example, I'm not sure what you mean here.

> Shortcodes: It's not clear if callbacks are supported here as well.

Could you elaborate?

---

Overall, big 👍 for the review and comments.

**Comment by borekb** [comment]
As for your feature suggestions:

> Add a filter to determine the plugin descriptors directory. As a plugin developer I would prefer to bury them deeper within the plugin's directory structure and avoid having a hidden directory there.

Let's wait if we get more feedback like this. I think it's beneficial if VP definitions are always found at the same, predictable place.

> Ignore git repositories when VersionPress is activated (and later on). Sure, it is possible to achieve this manually but I have to admint it's quite annoying.

Do you mean submodules? Git repos cloned to child folders?

> The possibility to favourite certain commits and then display only those favourites would be great. Use cases:
    - prototyping and frequently coming back to one "starting point"
    - repeatedly debugging some complex operation that involves database updates

Yep, very much on our roadmap :) See https://github.com/versionpress/versionpress/issues/1044.

**Comment by zaantar** [comment]
Hi there, I'm sorry for the delay.

> > Action detection: It's not entirely clear to me who (and when) should report them to VersionPress
>
> Seems pretty clear to me from the Actions detection section. What do you find unclear? Any suggestions of how to improve it?

For example, when my plugin is about to perform some non-database action - do I call `vp_force_action` before the action, or after it's finished? Or it's the same? Can I call it before the `init` action? Etc.

I would also maybe rephrase "Non-database actions are tracked manually (...)" to "Non-database actions can be tracked by manually calling a global `vp_force_action()` function."

> > VPID: Why is it needed? The reader (not too familiar with VersionPress) might get confused here, the difference between an ID and VPID is not really clear; a very brief explanation or a link to another document would help a lot here.
>
> I can see how the distinction between an ID and a VPID might be slightly confusing. An attempt to explain it is in the Identifying entities section, do you have any suggestions on how to make it better?

I think I must have missed the note referencing this section from "Files to commit with an action". After re-reading, I think it's good enough. 

> M:N references start with an example, I'm not sure what you mean here.

My bad, me neither anymore. All good there.

> > Shortcodes: It's not clear if callbacks are supported here as well.
>
> Could you elaborate?

For example, it's not clear if `shortcodes.yml` it is also possible to determine the entity type by providing a reference to a PHP function like this:

```
shortcodes:
  gallery:
    id: '@\MyPlugin\VersionPress::get_entity_type'
```

I think one sentence like "you can use the same types of reference as in actions.yml" would be enough.

> > Add a filter to determine the plugin descriptors directory. As a plugin developer I would prefer to bury them deeper within the plugin's directory structure and avoid having a hidden directory there.
>
> Let's wait if we get more feedback like this. I think it's beneficial if VP definitions are always found at the same, predictable place.

I agree that it's beneficial, but I fear that forcing everyone to use the hidden directory might complicate things. For example, would it be ok with the plugin review team on wordpress.org? Wouldn't it cause technical problems with plugin installation under some circumstances? (I'm just worried without any actual information to support this concern, but I would suggest to investigate about possible issues if you decide to keep the `.versionpress` directory name)

> > Ignore git repositories when VersionPress is activated (and later on). Sure, it is possible to achieve this manually but I have to admint it's quite annoying.
>
> Do you mean submodules? Git repos cloned to child folders?

Yes, exactly. This happens especially when using VersionPress on a dev site for developing a plugin (or a theme). And I don't think there's a reason for tracking submodules by default, when someone has them in their site, they would probably prefer VP to stay away from them. Or do you think there's a viable use case for this?

Thanks for your replies to all my remarks and feature requests. :+1:

**Comment by borekb** [comment]
> Shortcodes: It's not clear if callbacks are supported here as well.

Oh, got it now, function calls are not supported in `shortcodes.yml`. If they ever are, the document will mention that.

> `.versionpress` directory name, etc.

This might evolve if it causes issues, for sure.

> Nested Git repositories (submodules)

Agree, it's tracked here: https://github.com/versionpress/versionpress/issues/583
